### PR TITLE
[stdlib] Support reversed for Dict

### PR DESCRIPTION
**Pull request by jayzhan211**
Part of https://github.com/modularml/mojo/issues/2325

Follow up PR on #2327

I done the `values` and `items` of reversed dict, since #2327 is not yet merged, so this PR also includes #2327.
If this PR looks good, we can merge this, otherwise #2327 can be merged beforehand

### CODE REVIEW

**Comment by JoeLoser** [comment]
Feel free to rebase now that https://github.com/modularml/mojo/commit/abf2975d48fc7631e0e2ca741340289ec622f5d0 just landed for reversing the keys.  Sorry I didn't see this PR before merging the smaller one.

**Comment by jayzhan211** [comment]
@JoeLoser Done!

**Comment by jayzhan211** [comment]
I just notice Joe is on PTO, @ConnorGray ready for review!

**Comment by JoeLoser** [comment]
@jayzhan211 do you mind rebasing and fixing up the merge conflicts?  After that, happy to take a look.  Thank you!

**Comment by jayzhan211** [comment]
@JoeLoser Sure!

**Comment by jayzhan211** [comment]
@JoeLoser  Friendly ping, this PR is ready for review, thanks!

**Comment by JoeLoser** [comment]
> @JoeLoser Friendly ping, this PR is ready for review, thanks!

Thanks for the friendly ping.  I appreciate you rebasing — I just imported this PR!

File: stdlib/src/collections/dict.mojo [hunk 1]
```diff
@@ -177,15 +178,26 @@ struct _DictValueIter[
...
 
     alias ref_type = Reference[V, dict_mutability, dict_lifetime]
 
-    var iter: _DictEntryIter[K, V, dict_mutability, dict_lifetime]
+    var iter: _DictEntryIter[K, V, dict_mutability, dict_lifetime, forward]
 
     fn __iter__(self) -> Self:
         return self
 
+    fn __reversed__[
```
Comment (JoeLoser) [inline stdlib/src/collections/dict.mojo:191]: **Suggestion** Do you mind adding a changelog entry [here](https://github.com/modularml/mojo/blob/fd304d2f3f60a51816b8672cafdea41da35bd456/docs/changelog.md) for your work?  Adding an entry under the "New" section and crediting yourself and this PR link would be great!

File: stdlib/test/builtin/test_reversed.mojo [hunk 2]
```diff
@@ -55,29 +71,63 @@ def test_reversed_dict():
...
+
+    assert_equal(keys, "db")
+
+    # Refill dict
+    dict["c"] = 2
+    dict["a"] = 1
+    dict["b"] = 4
+    dict["d"] = 3
+
+    # dict: {'b': 4, 'd': 3, 'c': 2, 'a':1}
```
Comment (JoeLoser) [inline stdlib/test/builtin/test_reversed.mojo:95]:
```suggestion
    # dict: {'b': 4, 'd': 3, 'c': 2, 'a': 1}
```

File: stdlib/test/builtin/test_reversed.mojo [hunk 3]
```diff
@@ -55,29 +71,63 @@ def test_reversed_dict():
...
+
+    keys = String("")
+    check = 1
```
Comment (JoeLoser) [inline stdlib/test/builtin/test_reversed.mojo:98]: **Thought** Maybe soon we'll have an `enumerate` built-in function 😄

File: stdlib/test/builtin/test_reversed.mojo [hunk 4]
```diff
@@ -55,29 +71,63 @@ def test_reversed_dict():
...
+    check = 0
+    for val in reversed(empty_dict.values()):
+        # values is empty, should not reach here
+        check += 1
 
-    dict["d"] = 4
-    dict["a"] = 1
-    dict["b"] = 2
-    dict["e"] = 3
+    assert_equal(check, 0)
```
Comment (JoeLoser) [inline stdlib/test/builtin/test_reversed.mojo:121]: **Question** Perhaps it's easier to just `assert_equal(0, len(empty_dict.values()))`?  Or do you like the extra explicitness here when combined with reverse over the empty values?

**Review by JoeLoser — APPROVED** [review]
Looks great!  Do you mind adding a changelog entry, please?  I'm going to start suggesting external contributors can write their own changelog entries where it makes sense if you don't mind.  You're the first one I'm asking as we make this change 😄

File: stdlib/test/builtin/test_reversed.mojo (hunk 4 above)
Comment (jayzhan211) [reply stdlib/test/builtin/test_reversed.mojo:121]: Yes, I would prefer to test `reversed` with empty dict explicitly. `assert_equal(0, len(empty_dict.values()))` is not testing `reversed` to me

**Comment by jayzhan211** [comment]
@JoeLoser Done!

**Comment by JoeLoser** [comment]
✅🟣 This contribution has been merged 🟣✅

Your pull request has been merged to the internal upstream Mojo sources. It will be reflected here in the Mojo repository on the nightly branch during the next Mojo nightly release, typically within the next 24-48 hours.

We use Copybara to merge external contributions, click [here](https://github.com/google/copybara) to learn more.

**Comment by jayzhan211** [comment]
@JoeLoser Do you know why the change is not appeared in nightly?

https://github.com/modularml/mojo/blob/nightly/stdlib/src/builtin/reversed.mojo

**Comment by JoeLoser** [comment]
> @JoeLoser Do you know why the change is not appeared in nightly?
> 
> https://github.com/modularml/mojo/blob/nightly/stdlib/src/builtin/reversed.mojo

This is my fault, I'm sorry. Before we had this piece of automation up for auto commenting upon merging internally, I made a mistake here. This one has not landed internally and so isn't reflected on the nightly branch here. I've reopened this pull request and removed the incorrect labels I added.

I'll take a look on Monday for shepherding this PR through internally so it can land. I'm sorry about the troubles here - it's completely my fault. Thank you for calling it out!

**Comment by JoeLoser** [comment]
@jayzhan211 okay I actually merged it internally this time just now 🎉

**Pull request closed without merging**